components/tidb_query/src/rpn_expr/impl_arithmetic.rs

```diff
@@ -498,21 +484,16 @@ impl ArithmeticOpWithCtx for DecimalDivide {
     type T = Decimal;
 
     fn calc(ctx: &mut EvalContext, lhs: &Decimal, rhs: &Decimal) -> Result<Option<Decimal>> {
-        use crate::codec::mysql::Res;
-
-        Ok(match lhs / rhs {
-            Some(value) => match value {
-                Res::Ok(value) => Some(value),
-                Res::Truncated(_) => ctx.handle_truncate(true).map(|_| None)?,
-                Res::Overflow(_) => ctx
-                    .handle_overflow_err(Error::overflow(
-                        "DECIMAL",
-                        &format!("({} / {})", lhs, rhs),
-                    ))
-                    .map(|_| None)?,
-            },
-            None => ctx.handle_division_by_zero().map(|_| None)?,
-        })
+        Ok(if let Some(value) = lhs / rhs {
+            value
+                .into_result_with_overflow_err(
+                    ctx,
+                    Error::overflow("DECIMAL", format!("({} / {})", lhs, rhs)),
+                )
+                .map(Some)
+        } else {
+            ctx.handle_division_by_zero().map(|_| None)
+        }?)
```

Comment:
```go
func (s *builtinArithmeticDivideDecimalSig) evalDecimal(row chunk.Row) (*types.MyDecimal, bool, error) {
	a, isNull, err := s.args[0].EvalDecimal(s.ctx, row)
	if isNull || err != nil {
		return nil, isNull, err
	}

	b, isNull, err := s.args[1].EvalDecimal(s.ctx, row)
	if isNull || err != nil {
		return nil, isNull, err
	}

	c := &types.MyDecimal{}
	err = types.DecimalDiv(a, b, c, types.DivFracIncr)
	if err == types.ErrDivByZero {
		return c, true, handleDivisionByZeroError(s.ctx)
	} else if err == nil {
		_, frac := c.PrecisionAndFrac()
		if frac < s.baseBuiltinFunc.tp.Decimal {
			err = c.Round(c, s.baseBuiltinFunc.tp.Decimal, types.ModeHalfEven)
		}
	}
	return c, false, err
}
```

Reply:
The err handle is not same as tidb

Reply:
This is a bug of all functions in this module, I can't solve it in just one PR. The PR aims to handle the error with the best effort via `EvalContext`. For now, it's unreasonable to solve the inherent bugs in such a context.